### PR TITLE
plan: move some initialization out of convert2PhysicalPlan process

### PR DESCRIPTION
**Pull request by winoros**
We move some attributes' initialization out of convert2PhysicalPlan processs since convert2PhysicalPlan contains too many things and it's easy to forget the initialization of attributes and caused problems such as PR https://github.com/pingcap/tidb/pull/2140
Add `SetCorrelated` interface.
Fix  the wrong value of `correlated` of Join
PTAL @shenli @coocood @hanfei1991 @XuHuaiyu

### CODE REVIEW

**Comment by hanfei1991** [comment]
Do you need to add extractCorColumns for Physical***Join ?

File: plan/plan.go [hunk 1]
```diff
@@ -256,10 +258,10 @@ func (p *baseLogicalPlan) PredicatePushDown(predicates []expression.Expression)
 	return nil, p.self, nil
 }
 
-func (p *baseLogicalPlan) extractCorrelatedCols() []*expression.CorrelatedColumn {
+func (p *basePlan) extractCorrelatedCols() []*expression.CorrelatedColumn {
 	var corCols []*expression.CorrelatedColumn
 	for _, child := range p.children {
-		corCols = append(corCols, child.(LogicalPlan).extractCorrelatedCols()...)
+		corCols = append(corCols, child.(Plan).extractCorrelatedCols()...)
```
Comment (hanfei1991) [inline plan/plan.go]: child itself is already a Plan interface.

File: plan/physical_initialize.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+		}
+	}
+}
+
+// addCachePlan will add a Cache plan above the plan whose father's IsCorrelated() is true but its own IsCorrelated() is false.
+func addCachePlan(p PhysicalPlan, allocator *idAllocator) PhysicalPlan {
+	if len(p.GetChildren()) == 0 {
+		return p
+	}
+	np := p
```
Comment (hanfei1991) [inline plan/physical_initialize.go]: It seems we don't need a np. It's ok to use p directly.

**Comment by winoros** [comment]
yes, it's needed to implement extractCorColumns for Physical***Join

File: plan/plan.go [hunk 3]
```diff
@@ -98,12 +98,16 @@ type Plan interface {
 	GetID() string
 	// Check whether this plan is correlated or not.
 	IsCorrelated() bool
+	// Set the value of correlated
```
Comment (zimulala) [inline plan/plan.go]: Add . at end of a line

File: plan/physical_initialize.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+type physicalInitializer struct {
+	ctx       context.Context
+	allocator *idAllocator
+}
+
+func (ps *physicalInitializer) initialize(p PhysicalPlan) {
+	for _, child := range p.GetChildren() {
+		ps.initialize(child.(PhysicalPlan))
+
+		// init the attribute that will be affected by its children.
```
Comment (zimulala) [inline plan/physical_initialize.go]: s/init/Initialize ?

File: plan/physical_plan_test.go [hunk 5]
```diff
@@ -596,7 +596,7 @@ func (s *testPlanSuite) TestProjectionElimination(c *C) {
 		},
 		{
 			sql: "select t1.a from t t1 where t1.a in (select t2.a from t t2 where t1.a > 1)",
-			ans: "Apply{Table(t)->Table(t)->Cache->Selection}->Selection->Projection",
+			ans: "Apply{Table(t)->Table(t)->Selection}->Selection->Projection",
```
Comment (hanfei1991) [inline plan/physical_plan_test.go]: Why cache is cancelled here?

File: plan/physical_plans.go [hunk 6]
```diff
@@ -400,6 +400,43 @@ type Cache struct {
 	basePlan
 }
 
+func (p *PhysicalHashJoin) extractCorrelatedCols() []*expression.CorrelatedColumn {
+	corCols := p.basePlan.extractCorrelatedCols()
+	for _, fun := range p.EqualConditions {
```
Comment (hanfei1991) [inline plan/physical_plans.go]: Should we extract cols for Left/Right/Other Conditions?

File: plan/physical_plan_test.go (hunk 5 above)
Comment (winoros) [inline plan/physical_plan_test.go]: Currently physical inititialize and addCache is called after the ProjectionElimination, so cache doesn't appear in the TestProjectionElimination.

File: plan/physical_plans.go (hunk 6 above)
Comment (winoros) [inline plan/physical_plans.go]: yes. i missed them.

File: plan/physical_plans.go [hunk 7]
```diff
@@ -405,6 +405,15 @@ func (p *PhysicalHashJoin) extractCorrelatedCols() []*expression.CorrelatedColum
 	for _, fun := range p.EqualConditions {
 		corCols = append(corCols, extractCorColumns(fun)...)
 	}
+	for _, fun := range p.LeftConditions {
```
Comment (hanfei1991) [inline plan/physical_plans.go]: Add test case for this

File: plan/physical_initialize.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+func (ps *physicalInitializer) initialize(p PhysicalPlan) {
+	for _, child := range p.GetChildren() {
+		ps.initialize(child.(PhysicalPlan))
+
+		// Initialize the attribute that will be affected by its children.
+		p.SetCorrelated(p.IsCorrelated() || child.IsCorrelated())
+	}
+	switch pp := p.(type) {
+	case *PhysicalAggregation:
```
Comment (hanfei1991) [inline plan/physical_initialize.go]: Why Aggregation and Apply need to calculate "correlated" again?

Comment (winoros) [inline plan/physical_initialize.go]: In convert2PhysicalPlan, `correlated` doesn't initialize. So before physicalInitializer.initialize, the `correlated` is false.  The `correlated` of apply and aggregation is affected by itself, so i do it here.

Comment (hanfei1991) [inline plan/physical_initialize.go]: I mean the Logical Opt Phase has set the agg's correlation. If you need to set it here, remove the set correlated logic in aggregation push down and selection push down. And then consider all kinds of operators to set correlation. You can abstract it in interface.

File: plan/logical_plans.go [hunk 9]
```diff
@@ -112,6 +137,19 @@ func (p *Aggregation) extractCorrelatedCols() []*expression.CorrelatedColumn {
 	return corCols
 }
 
+// SetCorrelated implements Plan SetCorrelated interface.
```
Comment (hanfei1991) [inline plan/logical_plans.go]: SetCorrelated implements Plan interface. same as the followings.

**Comment by hanfei1991** [comment]
rest LGTM @coocood @shenli @zimulala PTAL

**Comment by winoros** [comment]
@coocood @shenli @zimulala PTAL

**Comment by hanfei1991** [comment]
LGTM

**Comment by hanfei1991** [comment]
@winoros Fix CI.

File: plan/plan.go [hunk 10]
```diff
@@ -98,12 +98,16 @@ type Plan interface {
 	GetID() string
 	// Check whether this plan is correlated or not.
 	IsCorrelated() bool
+	// Set the value of correlated.
```
Comment (coocood) [inline plan/plan.go]: This method needs more comment.
I checked all the implementations, the rule seems like below

1. if any child plan is correlated this plan is correlated, except for Apply.
2. Apply is correlated if first child is correlated, or second child has unknown correlated column.
3, If any expression in this plan is correlated, the plan is correlated.

If this rule is correct, we can add it in comment to make it easy to read.

Another thing to comment is when this function is called.
It seem like for any plan, this method is called as the final step of building when all the children and expressions have been set.

File: plan/logical_plan_builder.go [hunk 11]
```diff
@@ -93,6 +88,7 @@ func (b *planBuilder) buildAggregation(p LogicalPlan, aggFuncList []*ast.Aggrega
 	agg.GroupByItems = gbyItems
 	agg.SetSchema(schema)
 	agg.collectGroupByColumns()
+	agg.SetCorrelated()
```
Comment (coocood) [inline plan/logical_plan_builder.go]: All the logical plans calls `SetCorrelated` at the end, can we just use a function to recursively set Correlated of children, just like the physical plan initializer?

File: plan/physical_plan_builder.go [hunk 12]
```diff
@@ -1176,3 +1120,43 @@ func (p *Distinct) convert2PhysicalPlan(prop *requiredProperty) (*physicalPlanIn
 	p.storePlanInfo(prop, info)
 	return info, nil
 }
+
+// PhysicalInitializer will do the initialization of some attributes after the convert2Physical process finish.
+type physicalInitializer struct {
+	ctx       context.Context
```
Comment (coocood) [inline plan/physical_plan_builder.go]: If change the input to `Plan` interface, we can use this to set correlated for logical plans.

File: plan/logical_plan_builder.go (hunk 11 above)
Comment (hanfei1991) [inline plan/logical_plan_builder.go]: We can't, because we need to decide to use semi-join or not during plan building.

**Comment by coocood** [comment]
LGTM

**Pull request merged**